### ISSUE TITLE
paddle.autograd.grad not working

### ISSUE THREAD

**Issue by leonleeldc** [PaddlePaddle/Paddle#37688]
Hey, I am working to convert a torch work to paddle. I need to use paddle.autograd.grad to update gradients against some variables. However, it is not working while it is working for another group of vectors. 

The code snippet is as following, 

                for i in range(option['g_l_steps']):
                    gen_res = model(images, z, depths)
                    g_log_lkhd = 1.0 / (2.0 * option['g_llhd_sigma'] * option['g_llhd_sigma']) * mse_loss(
                        F.sigmoid(gen_res[0]), gts)
                    en = ebm_model(z)
                    #z_grad_g = g_log_lkhd.grad[0]
                    z_grad_g = g_log_lkhd.grad(z)
                    z_grad_g = paddle.autograd.grad(g_log_lkhd, z)[0]
                    z_grad_e = paddle.autograd.grad(en.sum(), z)[0]

                    z.data = z.data - 0.5 * option['g_l_step_size'] * option['g_l_step_size'] * (
                                z_grad_g + z_grad_e + 1.0 / (option['e_prior_sig'] * option['e_prior_sig']) * z.data)
                    z.data += option['g_l_step_size'] * fluid.layers.gaussian_random_batch_size_like(z, z.shape).data

May you help？
Thanks,

**Comment by yaoxuefeng6** [comment]
Codes looks good. Which groups not work in this case? Could you provide some runnable demo for us.

**Comment by leonleeldc** [comment]
Sure, let me share more details here. 

Looks that there are some problems in gen_res = model(images, z, depths).

In the model's forward function, we have the following, 


    def forward(self, x, z, depth=None):
        # print(z.size())
        #x = paddle.to_tensor(x.cpu().numpy())

        #x.stop_gradient = True
        #if depth != None:
            #depth = paddle.to_tensor(depth.cpu().numpy())
        x = paddle.concat((x,depth),1)
        x = self.conv_depth(x)
        features = self.encoder(x)

        x1 = features[-5]
        x2 = features[-4]
        x3 = features[-3]
        x4 = features[-2]
        x5 = features[-1]

        x1 = self.conv1(x1)
        x2 = self.conv2(x2)
        x3 = self.conv3(x3)
        x4 = self.conv4(x4)
        x5 = self.conv5(x5)

        z_noise = z
        z_noise = self.tile(z_noise, 2, x4.shape[self.spatial_axes[0]])
        z_noise = self.tile(z_noise, 3, x4.shape[self.spatial_axes[1]])

        x5 = paddle.concat((x5, z_noise), 1)
        x5 = self.noise_conv(x5)

        pred = self.sal_dec(x1,x2,x3,x4,x5)

        return [pred]

self.tile is defined as,
    def tile(self, a, dim, n_tile):
        """
        This function is taken form PyTorch forum and mimics the behavior of tf.tile.
        Source: https://discuss.pytorch.org/t/how-to-tile-a-tensor/13853/3
        """
        init_dim = a.shape[dim]
        repeat_idx = [1] * a.dim()
        repeat_idx[dim] = n_tile
        a = paddle.tile(a, repeat_times=repeat_idx)
        #a = paddle.to_tensor(np.repeat(a, repeat_idx[dim], axis=dim))
        order_index = paddle.to_tensor(np.concatenate([init_dim * np.arange(n_tile) + i for i in range(init_dim)]))
        return paddle.index_select(a, index=order_index, axis=dim)

That is where images and z interact. The torch code is very similar to this. For the running, it is as following, 

  import imp
Learning Rate: 1.00e+00
EBM Learning Rate: 1.00e+00
Epoch[001/100]:   0%|                                   | 0/219 [00:00<?, ?it/s]/home/dingcheng/dingcheng_env2/lib/python3.8/site-packages/paddle/nn/layer/norm.py:640: UserWarning: When training, we now always track global mean and variance.
  warnings.warn(

It goes for ever. 

I checked z_grad = paddle.autograd.grad(en.sum(), z)[0]. There is no problem for this line at all. 

Thanks,

**Comment by leonleeldc** [comment]
For the torch code, it looks as following, 

                for i in range(option['g_l_steps']):
                    gen_res = model(images, z, depths)
                    g_log_lkhd = 1.0 / (2.0 * option['g_llhd_sigma'] * option['g_llhd_sigma']) * mse_loss(
                        torch.sigmoid(gen_res[0]), gts)
                    en = ebm_model(z)
                    z_grad_g = torch.autograd.grad(g_log_lkhd, z)[0]


                    z_grad_e = torch.autograd.grad(en.sum(), z)[0]

                    z.data = z.data - 0.5 * option['g_l_step_size'] * option['g_l_step_size'] * (
                                z_grad_g + z_grad_e + 1.0 / (option['e_prior_sig'] * option['e_prior_sig']) * z.data)
                    z.data += option['g_l_step_size'] * torch.randn_like(z).data

**Comment by leonleeldc** [comment]
Now, I think that it should be related to the pretrained model loading. The following model is loaded from a converted torch model.
        model_state = paddle.load('./Transformer_based/RGBD_SOD_paddle/swin_base_patch4_window12_384.pdparams')
        self.encoder.set_dict(model_state)

But not sure what exact wrong here?
May you have some suggestions on what problems I may encounter here?
